docs/pages/reference.mdx

```diff
@@ -35,3 +35,58 @@
   ```
   pnpm mud set-version --commit <hash> && pnpm install
   ```
+
+## Debugging
+
+### Onchain components
+
+#### Foundry Debugging
+
+The MUD onchain components typically use the [Foundry development stack](https://book.getfoundry.sh/), meaning any of Foundry's debugging tools are available.
+Use this prodecure to view the output from [`anvil`](https://book.getfoundry.sh/reference/anvil/):
+
+1. Add `console.log` to your systems, edit their Solidity files.
+
+   - Import console.log.
+
+     ```solidity
+     import { console } from "forge-std/console.sol";
+     ```
+
+   - Add `console.log` statements inside your Solidity functions.
+
+     ```solidity
+     console.log("newValue:", newValue);
+     ```
+
+1. Run `anvil` manually.
+
+   ```sh
+   anvil -b 1 --block-base-fee-per-gas 0 | tee anvil.log
+   ```
+
+   You do not have to use `tee anvil.log`, but `anvil` logs so much that it is often useful to be able to look for information in the log file after it has scrolled off the screen.
+
+1. Edit `packages/contracts/package.json` to change the `scripts.dev` definition.
+
+   ```
+   "dev": "pnpm mud dev-contracts --tsgenOutput ../client/src/mud --rpc http://127.0.0.1:8545",
+   ```
+
+   Use `127.0.0.1`, there is a weird bug that makes it a problem to use `localhost` is some circumstances.
+
+1. Run `pnpm dev` normally.
+
+   Note that because of the way things work, there might be multiple copies of the same `console.log` message.
```

Comment:
Can you clarify this?

Reply:
No, I can't :-(. Do you know why that happens?

Reply:
On my system (`vanilla` template, `anvil` version 0.1.0), it produces one line for the original `increment` call, and then 20-21 on subsequent ones (the ones that come from the UI). I don't know what causes this, and I don't know enough to research. Somebody should do it when there's time, but I think we should already publish because it's a useful tool.

Reply:
I will prettify the language a bit and add a suggestion to run with `uniq`.

Reply:
I add language suggesting they use `uniq`, which reduces the issue.

Reply:
I verified the problem still happens with the latest (`main`) version. It looks like our code calls `IncrementSystem.increment` multiple times in calls that get reverted.

This is ugly, but I think it would be best to publish this technique already, and fix things when we have time to figure it out. It is already a useful technique.